lib/metadata/schema-parser.js

```diff
@@ -941,6 +1004,79 @@ SchemaParserV2.prototype._parseUdt = function (udtInfo, row, callback) {
   });
 };
 
+/**
+ * Used to parse schema information for Cassandra versions 4.x and above.
+ *
+ * This parser similar to [SchemaParserV2] expect it also parses virtual
+ * keyspaces.
+ *
+ * @param {ControlConnection} cc The control connection to be used
+ * @param {Function} udtResolver The function to be used to retrieve the udts.
+ * @constructor
+ * @ignore
+ */
+function SchemaParserV3(cc, udtResolver) {
+  SchemaParserV2.call(this, cc, udtResolver);
+  this.selectVirtualTable = _selectVirtualTable;
+  this.selectVirtualColumns = _selectVirtualColumns;
+}
+
+util.inherits(SchemaParserV3, SchemaParserV2);
+
+/** @override */
+SchemaParserV3.prototype.getKeyspaces = function (waitReconnect, callback) {
+  const self = this;
+  const keyspaces = {};
+  const queries = [
+    { query: _selectAllKeyspacesV2, virtual: false },
+    { query: _selectAllVirtualKeyspaces, virtual: true }
+  ];
+  utils.each(queries, (q, cb) => {
+    self.cc.query(q.query, waitReconnect, (err, result) => {
+      if (err) {
+        return cb(err);
+      }
+      for (let i = 0; i < result.rows.length; i++) {
+        const ksInfo = self._parseKeyspace(result.rows[i], q.virtual);
+        keyspaces[ksInfo.name] = ksInfo;
+      }
+      cb();
+    });
+  }, (err) => {
+    callback(err, keyspaces);
+  });
+};
+
+/** @override */
+SchemaParserV3.prototype.getKeyspace = function (name, callback) {
+  const self = this;
+  this.cc.query(util.format(_selectSingleKeyspaceV2, name), function (err, result) {
+    if (err) {
+      return callback(err);
+    }
+    const row = result.rows[0];
+    if (!row) {
+      // if not found try virtual keyspace
+      return self._getVirtualKeyspace(name, callback);
+    }
+    callback(null, self._parseKeyspace(row, false));
+  });
+};
+
+SchemaParserV3.prototype._getVirtualKeyspace = function (name, callback) {
```

Comment:
We could simplify this and the above logic into a single method that receives the query and whether or not it's virtual.

Reply:
To clarify, I meant using an instance method that calls `cc.query()` and parses the keyspace, in order to make `getKeyspace()` method be 2 nested calls:

```javascript
SchemaParserV3.prototype.getKeyspace = function (name, callback) {
  // Get normal (non-virtual) keyspace
  this._getKeyspace(_selectSingleKeyspaceV2, name, false, (err, ks) => {
    if (err) return callback(err);
    if (!ks) {
      // Try to get virtual ks
      return this._getKeyspace(_selectSingleKeyspaceV2, name, true, callback);
    }
    return callback(err, ks);
  });
};
```

Reply:
Ah, I see.  I felt that there was something I was missing when I made that change.  I'll change it to your suggestion since that makes sense to me :+1: